python/lsst/daf/butler/datastores/genericDatastore.py

```diff
@@ -108,19 +108,37 @@ def _register_datasets(self, refsAndInfos):
         self.registry.insertDatasetLocations(self.name, expandedRefs)
         self.addStoredItemInfo(expandedRefs, expandedItemInfos)
 
-    def _remove_from_registry(self, ref):
-        """Remove rows from registry.
+    def _move_to_trash_in_registry(self, ref):
+        """Tell registry that this dataset and associated components
+        are to be trashed.
 
         Parameters
         ----------
         ref : `DatasetRef`
-            Dataset to remove from registry.
+            Dataset to mark for removal from registry.
+
+        Notes
+        -----
+        Dataset is not removed from internal stored item info table.
         """
-        self.removeStoredItemInfo(ref)
-        self.registry.removeDatasetLocation(self.name, ref)
-        for compRef in ref.components.values():
-            self.registry.removeDatasetLocation(self.name, compRef)
-            self.removeStoredItemInfo(compRef)
+
+        # Note that a ref can point to component dataset refs that
+        # have been deleted already from registry but are still in
+        # the python object. moveDatasetLocationToTrash will deal with that.
+        allRefs = [ref] + [compRef for compRef in ref.components.values()]
+        self.registry.moveDatasetLocationToTrash(self.name, allRefs)
+
+    def _remove_trashed_registry_datasets(self, refs):
+        """Remove the dataset entries that have been trashed
+
+        Parameters
+        ----------
+        refs : iterable of `DatasetRef`
+           Datasets to be removed. These datasets must be present in the trash.
+        """
+        for ref in refs:
+            self.removeStoredItemInfo(ref)
+        self.registry.emptyDatasetLocationsTrash(self.name, refs)
```

Comment:
Is `_remove_trashed_registry_datasets` used anywhere?

Reply:
No. It was an early implementation. Well spotted.